pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -41,6 +47,18 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	// Watch for changes to secondary resource
+	enqueueRequestForOwner := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &toolchainv1alpha1.UserAccount{},
+	}
+	if err := c.Watch(&source.Kind{Type: &userv1.User{}}, enqueueRequestForOwner); err != nil {
+		return err
+	}
+	if err := c.Watch(&source.Kind{Type: &userv1.Identity{}}, enqueueRequestForOwner); err != nil {
+		return err
+	}
+
```

Comment:
adding `UserIdentityMapping` watch NOT possible.  Basically, `UserIdentityMapping` Get() operation is NOT available and adding watch calls it and failed with error - no kind for `UserIdentityMappingList`

Reply:
The `UserIdentityMapping` Get() operation is available: https://docs.openshift.com/container-platform/3.11/rest_api/apis-user.openshift.io/v1.UserIdentityMapping.html but it's not possible to list/watch this kind of resources. But this is a detail - the result is the same: it's not possible to watch them. This means that it needs to be created at the same time when a User is created and then, in the next reconcile, check that the specific mapping exists.

Reply:
Yeah..
`UserIdentityMapping` is a virtual resource. It doesn't exist in etcd.
The `create/delete` rest endpoint for this resource just updates the corresponding User and Identity: https://github.com/openshift/openshift-apiserver/blob/27f8efac5366bf8ec380644e326a2212fc0ff6ea///pkg/user/apiserver/registry/useridentitymapping/rest.go#L83
And `get` generates it's on fly out of the corresponding User and Identity if they have references to each other - https://github.com/openshift/openshift-apiserver/blob/27f8efac5366bf8ec380644e326a2212fc0ff6ea/////pkg/user/apiserver/registry/useridentitymapping/rest.go#L313:6

So, there is nothing to watch. We should watch User and Identity instead. And create them. Also in every reconcile if both User and Identity exist we should try to get UserIdentiyMapping (the api server will just check if User/Identity resources refer to each other). If it doesn't "exist" then "create" it. Please note that we should "create" it only if both User and Identity exist.
That "create" will update the corresponding User/Identity which will trigger another reconcile. So finally we will get an event when everything is ready.

Reply:
@MatousJobanek @alexeykazakov thanks for details.

.. but neither `Watch()` nor `Get()` possible for `UserIdentityMapping` (though there is GET REST API for the same)

Below Get call will give error:
**Code:** `r.client.Get(context.TODO(), types.NamespacedName{Name: name}, mapping)`
**Error:** no kind "UserIdentityMappingList" is registered for version "user.openshift.io/v1" in scheme "pkg/runtime/scheme.go:101"

Overall, I have already done what @alexeykazakov mentioned.  Plz check `encureMapping()` [link](https://github.com/codeready-toolchain/member-operator/blob/406473b5aed387786d343093ddbe3b27911d6224/pkg/controller/useraccount/useraccount_controller.go#L158).

Now, I am not getting what is missing and what should I change ??

Reply:
It looks good now. Just need to update the status.